linmod/data.py

```diff
@@ -289,16 +289,33 @@ def main(cfg: Optional[dict]):
             .then(pl.col("lineage"))
             .otherwise(pl.lit("other"))
         )
+        .collect()
+    )
+
+    # Generate every combination of date-division-lineage, so that:
+    #  1. The evaluation dataset will be evaluation-ready, with 0 counts
+    #     where applicable
+    #  2. The modeling dataset will have every lineage of interest represented,
+    #     even if a lineage was only sampled in the evaluation period
+    observations_key = expand_grid(
+        date=full_df["date"].unique(),
+        division=full_df["division"].unique(),
+        lineage=full_df["lineage"].unique(),
     )
 
     eval_df = (
         full_df.group_by("lineage", "date", "division")
         .agg(count=pl.len())
+        .join(
+            observations_key,
+            on=("date", "division", "lineage"),
+            how="right",
+        )
         .with_columns(
-            fd_offset=(pl.col("date") - forecast_date).dt.total_days()
+            fd_offset=(pl.col("date") - forecast_date).dt.total_days(),
+            count=pl.col("count").fill_null(0),
```

Comment:
fill null always makes me nervous

are we sure that all the nulls are because of expected things? is it work checking that `full_df` doesn't already have nulls in it that are getting inadvertently filled with zeros?

Reply:
Good point. What should we do if the dataset has a null count, though? Drop the row?

Reply:
or just error out

Reply:
Yes yes, I forgot that `count` is something we create, so this code does feel rather safe to me. That being said, I'll add null checks to the eval/modeling dataframes immediately before exporting